tests.py

```diff
@@ -17,172 +23,183 @@
 
-    mha = MultiHeadAttention(num_heads, d_model)
-    output = mha(query, key, value, mask)
-    
-    assert output.shape == (batch_size, seq_length, d_model)
-
-def test_positionwise_ffn():
-    d_ff = 32
-    d_model = 16
-    seq_length = 10
-    batch_size = 2
-    dropout = 0.1
-    ffn = PositionwiseFFN(d_ff, d_model, dropout)
-    x = torch.rand(batch_size, seq_length, d_model)
-    
-    out = ffn(x)
-    assert out.shape == (batch_size, seq_length, d_model)
-
-def test_encoder_layer():
-    num_heads = 4
-    d_model = 16
-    d_ff = 32
-    seq_length = 10
-    batch_size = 2
-    dropout = 0.1
-    x = torch.rand(batch_size, seq_length, d_model)
-    mask = torch.ones(batch_size, num_heads, seq_length, seq_length)
+    torch.manual_seed(0) # adding this for reproducibility to check if outputs are also close
+    my_mha = MultiHeadAttention(num_heads=num_heads, d_model=d_model)
+    my_ouput = my_mha(query, key, value)
 
-    encoder_layer = EncoderLayer(num_heads, d_model, d_ff, dropout)
-    out = encoder_layer(x, mask)
-    
-    assert out.shape == (batch_size, seq_length, d_model)
-
-def test_encoder():
-    num_blocks = 2
-    num_heads = 4
-    d_model = 16
-    d_ff = 32
-    seq_length = 10
-    batch_size = 2
-    dropout = 0.1
-    x = torch.rand(batch_size, seq_length, d_model)
-    mask = torch.ones(batch_size, num_heads, seq_length, seq_length)
+    torch.manual_seed(0)
+    torch_mha = nn.MultiheadAttention(d_model, num_heads)
+    torch_output, torch_attention_output_weights = torch_mha(query, key, value)
 
-    encoder = Encoder(num_blocks, num_heads, d_model, d_ff, dropout)
-    out = encoder(x, mask)
-    
-    assert out.shape == (batch_size, seq_length, d_model)
-
-def test_decoder_layer():
-    num_heads = 4
-    d_model = 16
-    d_ff = 32
-    seq_length = 10
-    batch_size = 2
-    dropout = 0.1
-    x = torch.rand(batch_size, seq_length, d_model)
+    assert my_ouput.shape == torch_output.shape, "MHA output doesn't have the same shape as PyTorch's implementation."
+
+    # couldn't get the reproducibility to work, will look further into this later. 
+    # assert torch.allclose(my_ouput, torch_output), "MHA output does not match PyTorch's implementation."
+
+@pytest.mark.parametrize("batch_size, seq_length, num_heads, d_ff, d_model, dropout", list(itertools.product(BATCH_SIZES, SEQ_LENGTHS, NUM_HEADS, D_FFS, EMB_SIZES, DROPOUTS)))
+def test_encoder_layer(batch_size, seq_length, num_heads, d_ff, d_model, dropout):
+
+    src = torch.rand(batch_size, seq_length, d_model)
+
+    my_encoder_layer = EncoderLayer(num_heads=num_heads, d_ff=d_ff, d_model=d_model, dropout=dropout)
+    torch_encoder_layer = nn.TransformerEncoderLayer(d_model, num_heads, d_ff, dropout, batch_first=True)
+
+    my_output = my_encoder_layer(src)
+    torch_output = torch_encoder_layer(src)
+
+    assert my_output.shape == torch_output.shape, "Encoder Layer shape doesn't match PyTorch's implementation."
```

Comment:
## Requested Change
* Assert they `my_output` matches the `torch_output`

Reply:
Do you mean input by input here? i.e. using something like 
`torch.allclose(my_output, torch_output, ...)`
I didn't do that in this case because we're randomly intializing two different layers, so we wouldn't expect the outputs to match. OR do you mean something else?

Reply:
Yeah, you need to use `torch.allclose(my_output, torch_output, ...)` here.
To make sure that the outputs match, you should add these two lines at the top of your file:
```py
import torch
import numpy as np

torch.manual_seed(1337)
np.random.seed(1337)
```
This will make sure that the random values will be the same